File: Stream_of_Characters.swift
/*
Implement the StreamChecker class as follows:

StreamChecker(words): Constructor, init the data structure with the given words.
query(letter): returns true if and only if for some k >= 1, the last k characters queried (in order from oldest to newest, including this letter just queried) spell one of the words in the given list.


Example:

StreamChecker streamChecker = new StreamChecker(["cd","f","kl"]); init the dictionary.
streamChecker.query('a');          return false
streamChecker.query('b');          return false
streamChecker.query('c');          return false
streamChecker.query('d');          return true, because 'cd' is in the wordlist
streamChecker.query('e');          return false
streamChecker.query('f');          return true, because 'f' is in the wordlist
streamChecker.query('g');          return false
streamChecker.query('h');          return false
streamChecker.query('i');          return false
streamChecker.query('j');          return false
streamChecker.query('k');          return false
streamChecker.query('l');          return true, because 'kl' is in the wordlist


Note:

1 <= words.length <= 2000
1 <= words[i].length <= 2000
Words will only consist of lowercase English letters.
Queries will only consist of lowercase English letters.
The number of queries is at most 40000.
*/

/*
Solution 1: Trie

Time complexity:
init() <- trie insert, O(mn) m is words[i] length, n is words length
query() <- trie search, O(k) k is current input length
Space complexity: O(mn) for build the trie
*/
class StreamChecker {
    var trie: Trie
    var input = [Character]() // save input as reverse order
    init(_ words: [String]) {
        trie = Trie(words)
    }

    func query(_ letter: Character) -> Bool {
        input.insert(letter, at: 0)
        return trie.search(input)
    }
}

class Trie {
    class TrieNode {
        var child = [Character: TrieNode]()
        var isWord = false
    }

    var root = TrieNode()

    init(_ words: [String]) {
        for word in words {
            insert(word)
        }
    }

    func insert(_ word: String) {
        var cur = root
        for c in word.reversed() {
            if cur.child[c] == nil {
                cur.child[c] = TrieNode()
            }
            cur = cur.child[c]!
        }
        cur.isWord = true
    }

    func search(_ input: [Character]) -> Bool {
        var cur = root
        for c in input {
            if cur.child[c] == nil {
                return false
            }
            cur = cur.child[c]!
            if cur.isWord {
                return true
            }
        }
        return false
    }
}

/**
 * Your StreamChecker object will be instantiated and called as such:
 * let obj = StreamChecker(words)
 * let ret_1: Bool = obj.query(letter)
 */

File: Stream_of_Characters.java
/*
Design an algorithm that accepts a stream of characters and checks if a suffix of these characters is a string of a given array of strings words.

For example, if words = ["abc", "xyz"] and the stream added the four characters (one by one) 'a', 'x', 'y', and 'z', your algorithm should detect that the suffix "xyz" of the characters "axyz" matches "xyz" from words.

Implement the StreamChecker class:

StreamChecker(String[] words) Initializes the object with the strings array words.
boolean query(char letter) Accepts a new character from the stream and returns true if any non-empty suffix from the stream forms a word that is in words.


Example 1:

Input
["StreamChecker", "query", "query", "query", "query", "query", "query", "query", "query", "query", "query", "query", "query"]
[[["cd", "f", "kl"]], ["a"], ["b"], ["c"], ["d"], ["e"], ["f"], ["g"], ["h"], ["i"], ["j"], ["k"], ["l"]]
Output
[null, false, false, false, true, false, true, false, false, false, false, false, true]

Explanation
StreamChecker streamChecker = new StreamChecker(["cd", "f", "kl"]);
streamChecker.query("a"); // return False
streamChecker.query("b"); // return False
streamChecker.query("c"); // return False
streamChecker.query("d"); // return True, because 'cd' is in the wordlist
streamChecker.query("e"); // return False
streamChecker.query("f"); // return True, because 'f' is in the wordlist
streamChecker.query("g"); // return False
streamChecker.query("h"); // return False
streamChecker.query("i"); // return False
streamChecker.query("j"); // return False
streamChecker.query("k"); // return False
streamChecker.query("l"); // return True, because 'kl' is in the wordlist


Constraints:

1 <= words.length <= 2000
1 <= words[i].length <= 2000
words[i] consists of lowercase English letters.
letter is a lowercase English letter.
At most 4 * 104 calls will be made to query.
 */

 /*
Solution 1: Trie

Time complexity:
init() <- trie insert, O(mn) m is words[i] length, n is words length
query() <- trie search, O(k) k is current input length
Space complexity: O(mn) for build the trie
*/
class StreamChecker {
    Trie trie;
    StringBuilder input;

    public StreamChecker(String[] words) {
        this.trie = new Trie(words);
        this.input = new StringBuilder();
    }

    public boolean query(char letter) {
        this.input.insert(0, letter);
        return trie.search(this.input);
    }
}

class Trie {
    class TrieNode {
        public Map<Character, TrieNode> child;
        public boolean isWord;

        public TrieNode() {
            this.child = new HashMap<>();
            this.isWord = false;
        }
    }

    public TrieNode root;

    public Trie(String[] words) {
        this.root = new TrieNode();
        for (int i = 0; i < words.length; i++) {
            insert(words[i]);
        }
    }

    public void insert(String word) {
        TrieNode cur = this.root;

        StringBuilder sb = new StringBuilder(word);
        sb.reverse();
        for(int i = 0; i < sb.length(); i++) {
            if(cur.child.get(sb.charAt(i)) == null) {
                cur.child.put(sb.charAt(i), new TrieNode());
            }
            cur = cur.child.get(sb.charAt(i));
        }
        cur.isWord = true;
    }

    public boolean search(StringBuilder input) {
        TrieNode cur = this.root;

        for(int i = 0; i < input.length(); i++) {
            if(cur.child.get(input.charAt(i)) == null) {
                return false;
            }
            cur = cur.child.get(input.charAt(i));
            if (cur.isWord) {
                return true;
            }
        }
        return false;
    }
}

/**
 * Your StreamChecker object will be instantiated and called as such:
 * StreamChecker obj = new StreamChecker(words);
 * boolean param_1 = obj.query(letter);
 */